src/dbnode/persist/fs/files.go

```diff
@@ -373,51 +415,136 @@ func (a fileSetFilesByTimeAndVolumeIndexAscending) Less(i, j int) bool {
 	return ti.Equal(tj) && ii < ij
 }
 
-func componentsAndTimeFromFileName(fname string) ([]string, time.Time, error) {
-	components := strings.Split(filepath.Base(fname), separator)
-	if len(components) < 3 {
-		return nil, timeZero, fmt.Errorf("unexpected file name %s", fname)
+// Returns the positions of filename delimiters ('-' and '.') and the number of
+// delimeters found, to be used in conjunction with the intComponentFromFilename
+// function to extract filename components. This function is deliberately
+// optimized for speed and lack of allocations, since filename parsing is known
+// to account for a significant proportion of system resources.
+func delimiterPositions(baseFilename string) ([maxDelimNum]int, int) {
+	var (
+		delimPos    [maxDelimNum]int
+		delimsFound int
+	)
+
+	for i := range baseFilename {
+		if r := baseFilename[i]; r == separatorRune || r == fileSuffixDelimeterRune {
+			delimPos[delimsFound] = i
+			delimsFound++
+
+			if delimsFound == len(delimPos) {
+				// Found the maximum expected number of separators.
+				break
+			}
+		}
 	}
-	str := strings.Replace(components[1], fileSuffix, "", 1)
-	nanoseconds, err := strconv.ParseInt(str, 10, 64)
+
+	return delimPos, delimsFound
+}
+
+// Returns the the specified component of a filename, given the positions of
+// delimeters. Our only use cases for this involve extracting numeric
+// components, so this function assumes this and returns the component as an
+// int64.
+func intComponentFromFilename(
+	baseFilename string,
+	componentPos int,
+	delimPos [maxDelimNum]int,
+) (int64, error) {
+	var start int
```

Comment:
can you do `var start = 0`, that would have given me a nudge in the right direction to understand how you're using the positions

Reply:
Yeah makes sense.